src/vsr/replica.zig

```diff
@@ -5051,19 +5049,40 @@ pub fn ReplicaType(
             return vsr.Checkpoint.trigger_for_checkpoint(self.op_checkpoint_next()).?;
         }
 
+        /// Returns checkpoint id associated with the op.
+        ///
+        /// Normally, this is just the id of the checkpoint the op builds on top. However, ops
+        /// between a checkpoint and its trigger can't know checkpoint's id yet, and instead use
+        /// the id of the previous checkpoint.
+        ///
+        /// Returns `null` for ops which are too far in the past/future to know their checkpoint
+        /// ids.
         fn checkpoint_id_for_op(self: *const Self, op: u64) ?u128 {
+            // Case 1: for the root op, checkpoint id is zero.
             if (op == 0) return Header.Prepare.root(self.cluster).checkpoint_id;
 
             if (self.op_checkpoint() > 0) {
-                if (op < self.op_repair_min()) return null;
-
-                const op_checkpoint_next_trigger_previous =
+                const op_checkpoint_trigger =
                     vsr.Checkpoint.trigger_for_checkpoint(self.op_checkpoint()).?;
-                if (op <= op_checkpoint_next_trigger_previous) {
+                if (op <= op_checkpoint_trigger) {
+                    if (op + constants.vsr_checkpoint_interval <= op_checkpoint_trigger) {
+                        // Case 2: op is from a too distant past for us to know its checkpoint id.
+                        return null;
+                    }
+                    // Case 3: op is from the previous checkpoint whose id we still remember.
                     return self.superblock.working.vsr_state.checkpoint.previous_checkpoint_id;
                 }
+
+                assert(op + constants.vsr_checkpoint_interval > self.op_checkpoint_next_trigger());
             }
-            return self.superblock.working.checkpoint_id();
```

Comment:
I think this wasn't entirely correct actually! For ops far in the future, we'd use current checkpoint ID, rather than `null`.